### ISSUE TITLE
Expand search by supporting for timeranges on CLOCK, SCHEDULED or DEADLINE

### ISSUE THREAD

**Issue by munen** [200ok-ch/organice#504]
This will also enable the [Clock display functionality](https://github.com/200ok-ch/organice/issues/494) to be filtered to specific time ranges.

If this feature will not be implemented, a less generic option is to implement Emacs Org mode `org-clock-display-default-range` in the settings.

**Comment by schoettl** [comment]
Interesting! To comply to the current search syntax, the new (how did I called it?) search terms would be

```
scheduled:2020-10-11    deadline:2020-10-11     clock:2020-10-11
```

We could allow abbreviations for those, e.g. `sched` and `dead[l]`.

Regarding to the date specification:

I have the feeling that @munen and I agree that mu4e query date ranges is the best ;) – right?

`man mu-query` -> `/^DATE RANG`

Are there other suggestions (e.g. [hledger date query syntax](https://hledger.org/hledger.html#period-expressions))?

**Comment by schoettl** [comment]
On that way we can also implement the full-text search with `description:word` or `desc:'word or text'`!

**Comment by munen** [comment]
> Interesting! To comply to the current search syntax, the new (how did I called it?)

I think you called it "headline filter". "search syntax" works for me, too. The latter is probably easier to find for a new contributor, but the former is technically equally correct.

> search terms would be
> `scheduled:2020-10-11    deadline:2020-10-11     clock:2020-10-11`

I was thinking the same :+1: 

> I have the feeling that @munen and I agree that mu4e query date ranges is the best ;) – right?

Full agreement :100: 

Here's the doc for mu4e queries: https://www.djcbsoftware.nl/code/mu/mu4e/Queries.html

The relevant terms would be:

- YYYY-MM-DD
- \d[dwmy] (i.e. `2d` for 2 days)
- today
- now

The separator between two dates would be `..`.

The second date is optional, so these would be valid terms:

- `scheduled: today`
- `scheduled: 1w..`

> On that way we can also implement the full-text search with description:word or desc:'word or text'!

That would certainly be hugely appreciated by the community :+1:

**Comment by tarnung** [comment]
> The relevant terms would be:
> 
>     * YYYY-MM-DD
> 
>     * \d[dwmy] (i.e. `2d` for 2 days)
> 
>     * today
> 
>     * now

I suggest to also allow
* YYYY
* YYYY-MM

> The separator between two dates would be `..`.
> 
> The second date is optional, so these would be valid terms:
> 
>     * `scheduled: today`
> 
>     * `scheduled: 1w..`

Do I understand it correctly that the eluded date in a range is assumed to be 'today'?
So '1w..' would mean 'from one week ago until today'. Shouldn't there also be the possibility to leave out the first date? So '..1w' would mean 'from today until one week in the future'.

I'm in favour of shortening the search terms to 'sched:' and 'dead:'. Maybe we should allow for both the abbreviation and the spelled out form.

Concerning the implementation, I wouldn't mind taking a first jab at it unless @schoettl calls dibs.

**Comment by schoettl** [comment]
> I suggest to also allow
> 
>     * YYYY
> 
>     * YYYY-MM

I agree. This is also allowed by the mu4e query filter syntax.


> I'm in favour of shortening the search terms to 'sched:' and 'dead:'. Maybe we should allow for both the abbreviation and the spelled out form.

I agree with the short terms but I'd also allow (and auto-complete) the full terms also because then it's more clear for newcomers. It should be trivial to allow both in the parser grammar.

> Concerning the implementation, I wouldn't mind taking a first jab at it unless @schoettl calls dibs.

@tarnung Feel free to do it. Have you worked with pegjs parser generator before? It's quite nice. The relevant files are:

- headline_filter_parser.grammar.pegjs
- headline_filter.js
- headline_filter.test.js

Starting point would probably be:
```
PlainTerm
  = TermText
  / TermProp
  / TermTag
  / TermField    <-- new, a "search term" to search in a specific field which can be TermDescription, TermScheduled, TermDeadline, ...
```

**Comment by schoettl** [comment]
> The second date is optional, so these would be valid terms:
> 
>     * `scheduled: today`
> 
>     * `scheduled: 1w..`

@munen Did you insert the space intentionally? I've never seen a space after the colon on GitHub issues or mu4e query – not sure if that works. To be honest, I would not allow the space. I think it's simpler and cleaner to implement in the grammar without it.

**Comment by munen** [comment]
> I agree with the short terms but I'd also allow (and auto-complete) the full terms also because then it's more clear for newcomers. It should be trivial to allow both in the parser grammar.

:+1: 

> @munen Did you insert the space intentionally? I've never seen a space after the colon on GitHub issues or mu4e query – not sure if that works. To be honest, I would not allow the space. I think it's simpler and cleaner to implement in the grammar without it.

Not intentionally, no. I've quickly typed this up on my media laptop (no mu4e there) in between Formula 1 sessions from the top of my head. Apologies for the mistake. I agree that the space should not be allowed (as is the case for `mu` as well).

**Comment by munen** [comment]
> Do I understand it correctly that the eluded date in a range is assumed to be 'today'?

The eludated date is `now`. `now` being the actual time right now and `today` being todays date.

NB: The implementation for those in `mu` is here: https://github.com/djcb/mu/blob/0f5973c61fdb39d1210ae5f750448e2dfdb1aa19/lib/utils/mu-utils.cc#L270-L294

The man page for the queries that shows better examples than the website that I initially posted is here: https://github.com/djcb/mu/blob/cc9816ba89188735a7a1d9b1f48a7d2ff683ed1f/man/mu-query.7#L180

From there, it's also clear that I didn't properly specify all units. To quote:

> Finally, you can use relative 'ago' times which express some time before now and consist of a number followed by a unit, with units *s* for seconds, *M* for minutes, *h* for hours, *d* for days, *w* for week, *m* for months and *y* for years.

> So '1w..' would mean 'from one week ago until today'.

Technically, it's 'from one week ago until now'. But in my understanding and in my test queries, that returns the same results.

> Shouldn't there also be the possibility to leave out the first date? So '..1w' would mean 'from today until one week in the future'.

Good point. In using `mu`, I never had the idea to try that, because `mu` is for searching mails(;

For `mu`, the search term `..today` gives me all mails. `..1d` gives me all emails from the past up to yesterday.

I assume this is where we could break with the semantics in `mu` and go with your proposal, because for Org this actually makes a lot of sense. @schoettl Do you agree?

**Comment by munen** [comment]
Next to `CLOCK`, `SCHEDULED` and `DEADLINE` timestamps, there's plain timestamps. Especially for calendars (aka agenda)

More information on how planning works in Org is in the 'sample' (https://github.com/200ok-ch/organice/blob/master/sample.org#planning), a regular appointment is [here](https://github.com/200ok-ch/organice/blob/master/sample.org#doctors-appointment-2020-10-07-wed) (note that the GH Org renderer omits `SCHEDULED` and `DEADLINE` information). The official Org mode documentation on that topic is [here](https://orgmode.org/manual/Deadlines-and-Scheduling.html#Deadlines-and-Scheduling) which also explains the rationale for plain timestamps.

Hence, I propose a `date:` qualifier that would work across all timestamps. That would also be in line with how the Emacs Org agenda works (limiting to a day or week will take into account `SCHEDULED`, `DEADLINE` and plain active timestamps).

**Comment by schoettl** [comment]
> I assume this is where we could break with the semantics in `mu` and go with your proposal, because for Org this actually makes a lot of sense. @schoettl Do you agree?

Yes, I like this semantic, it makes sense.

- `1w..xxx` = one week ago
- `xxx..1w` = one week in future

(where `xxx` is optional)

**Comment by tarnung** [comment]
I created a [pr](#504) with my first draft.
I wanted to allow as much as possible so there are some additions to the rules discussed above:

- `clock:..` searches for all items with time clocked
- `sched:w` searches for all scheduled items in the current week. (same for other units)
- `dead:..w` searches for all deadlines bewteen now and the end of the week.
- `sched:w..2m` if both ends of the range are relative, they refer back to the current moment. So this searches for all scheduled items between two months before now to the end of the current week.

Issues to discuss:

- Until now the filter terms only allow to specify days. With `now` and `h` as a time unit we have two elements that allow for higher precision in very limited cases. Is that the right feature set concerning precision? What schould be added or cut?

- Related to the issue above: `..1d` at the moment means 'from now till 24h from now'. It could also reasonably mean 'from now till tomorrow', meaning the end of that unit of time (the end of next day in this case). Which should it be?

- We have CLOCK items in logBookEntries and SCHEDULED, DEADLINE, TIMESTAMP_TITLE, TIMESTAMP_DESCRIPTION and TIMESTAMP_LOG_NOTES  in planningItems. Is there any other kind? And what does TIMESTAMP_LOG_NOTES refer to?

**Comment by schoettl** [comment]
> I created a [pr](#504) with my first draft.
> I wanted to allow as much as possible so there are some additions to the rules discussed above:
> 
>     * `clock:..` searches for all items with time clocked
> 
>     * `sched:w` searches for all scheduled items in the current week. (same for other units)
> 
>     * `dead:..w` searches for all deadlines between now and the end of the week.
> 
>     * `sched:w..2m` if both ends of the range are relative, they refer back to the current moment. So this searches for all scheduled items between two months before now to the end of the current week.

Wouldn't your description of the last one be `sched:2m..w`. The time range should generally be `start..end`.

> 
>     * Until now the filter terms only allow to specify days. With `now` and `h` as a time unit we have two elements that allow for higher precision in very limited cases. Is that the right feature set concerning precision? What schould be added or cut?
> 
>     * Related to the issue above: `..1d` at the moment means 'from now till 24h from now'. It could also reasonably mean 'from now till tomorrow', meaning the end of that unit of time (the end of next day in this case). Which should it be?

Maybe we can just stick to what `mu find` does. Maybe @munen can answer that; it's not clear from the mu-query man page, but we can just replicate mu behavior. (I don't have current mails in mu, so I can't test.)

> 
>     * We have CLOCK items in logBookEntries and SCHEDULED, DEADLINE, TIMESTAMP_TITLE, TIMESTAMP_DESCRIPTION and TIMESTAMP_LOG_NOTES  in planningItems. Is there any other kind? And what does TIMESTAMP_LOG_NOTES refer to?

I've added the log notes recently. You can see what it is when you press the "Note" icon (last icon in second row of the header toolbar). It's just an inactive timestamp in a org mode list below the headline.

I think your list of timestamps is complete.

But it's a good question: We can decide if we really want to search ALL timestamps or just active one. Hm, I don't have an opinion on that. I just think that including also inactive timestamps could sometimes yield too many results.

**Comment by tarnung** [comment]
> Wouldn't your description of the last one be sched:2m..w. The time range should generally be start..end.

Yes, edited.

I'll leave the discussion about active vs inactive timestamps to the seasoned org users.

**Comment by munen** [comment]
>  I wanted to allow as much as possible so there are some additions to the rules discussed above:

I've checked out your PR and tested the new rules. On a first run, I like all of them :partying_face: 

> Until now the filter terms only allow to specify days. With now and h as a time unit we have two elements that allow for higher precision in very limited cases. Is that the right feature set concerning precision? What schould be added or cut?

Yeah, I've added those to the issue, because `mu` supports them. Org also does support timestamps with hours and seconds. Having said that, in my years of using Emacs Org mode I _never_ had the inkling to filter with more time-specificity than a day. I'm not even aware that Emacs agendas have that ability built-in.

I vote to cut them. Apologies for briging it up.

> Related to the issue above: ..1d at the moment means 'from now till 24h from now'. It could also reasonably mean 'from now till tomorrow', meaning the end of that unit of time (the end of next day in this case). Which should it be?

If we want to want to keep with the semantics of `mu`, it's 'from now till 24h from now'. Here's an example of my emails showing this. You can see it when I go back 3d and 4d. Both return the 10th of October, but with 4d we get an AM date.

```
munen@lambda:~% date
Tue 13 Oct 2020 08:27:09 PM CEST
munen@lambda:~% mu find date:1d.. | head -n 1 | cut -d' ' -f 1-7
Mon 12 Oct 2020 09:03:14 PM CEST
munen@lambda:~% mu find date:2d.. | head -n 1 | cut -d' ' -f 1-7
Sun 11 Oct 2020 10:05:18 PM CEST
munen@lambda:~% mu find date:3d.. | head -n 1 | cut -d' ' -f 1-7
Sat 10 Oct 2020 10:00:48 PM CEST
munen@lambda:~% mu find date:4d.. | head -n 1 | cut -d' ' -f 1-7
Sat 10 Oct 2020 06:26:00 AM CEST
munen@lambda:~% mu find date:5d.. | head -n 1 | cut -d' ' -f 1-7
Thu 08 Oct 2020 09:06:27 PM CEST
```

I'm leaving the last point out as schoettl was faster in answering than me(;

In any case, I already want to say: Great job :100:

**Comment by munen** [comment]
> I'll leave the discussion about active vs inactive timestamps to the seasoned org users.

Strong vote to only use active timestamps for `scheduled`, `deadline` and regular timestamps. The reason to use inactive timestamps is to not have them affect the agenda. For logbook (clock) entries, it's different, because those are inactive (as they shouldn't contribute to the agenda). With the `clock:` search, you're exceeding standard Emacs Org mode search functionality imo(;

**Comment by tarnung** [comment]
Notes have inactive timestamps so those would not be searchable. But I guess you'd search for the content not the timestamp anyway.

**Comment by munen** [comment]
> Notes have inactive timestamps so those would not be searchable. But I guess you'd search for the content not the timestamp anyway.

Agreed. There's inactive timestamps for all kinds of things in Emacs. But making it inactive is an explicit statement that it should be left out of the agenda. 

A timestamp on a note is (per docstring in the code) the same as a state change note, i.e.

```
*** TODO Foo
    SCHEDULED: <2021-09-21 Tue +1y>
    :PROPERTIES:
    :LAST_REPEAT: [2020-10-13 Tue 10:28]
    :END:

    - State "DONE"       from "TODO"       [2020-10-13 Tue 10:28]
    - State "DONE"       from "TODO"       [2019-10-04 Fri 12:23]
    - State "DONE"       from "TODO"       [2018-11-18 Sun 15:43]
    - State "DONE"       from "TODO"       [2017-05-16 Tue 15:40]
    - State "DONE"       from "TODO"       [2016-04-22 Fri 19:47]
```

Here, I'd also definitively either search for "foo" or the scheduled date, not the past times when I've done the thing. If we include those, at some point, I'd find more or less all the headers for a specific time(;

**Comment by munen** [comment]
NB: Emacs has quite a few of these "state change" options, so some people might have _loads_ of inactive timestamps hanging around.

'logdone'	Record a timestamp when an item is marked as done.
'lognotedone'	Record timestamp and a note when DONE.
'logrepeat'	Record a time when reinstating a repeating item.
'lognoterepeat'	Record a note when reinstating a repeating item.
'lognoteclock-out'	Record a note when clocking out.
'logreschedule'	Record a timestamp when scheduling time changes.
'lognotereschedule'	Record a note when scheduling time changes.
'logredeadline'	Record a timestamp when deadline changes.
'lognoteredeadline'	Record a note when deadline changes.
'logrefile'	Record a timestamp when refiling.
'lognoterefile'	Record a note when refiling.

**Comment by munen** [comment]
Sweet, the clock display also already works in search:

![image](https://user-images.githubusercontent.com/505721/95903129-08deb280-0d96-11eb-9eed-62d99b0017c4.png)

It could use a bit of flexbox magic for alignment, but otherwise it looks good!

**Comment by tarnung** [comment]
I added the clock times to the TitleLine component without noticing that it is used to display items in search. I have noticed since though. But in search it should probably show a simple sum of the time tracked on this header, not a recursive sum. Which is probably another rare edge case since I suspect time is mostly tracked on leaf-nodes.

**Comment by munen** [comment]
> But in search it should probably show a simple sum of the time tracked on this header, not a recursive sum.

Curious, why do you say so? Let's say you're tracking time for a customer which has many projects (or epics) with many todos each. Then, you track (as you stated) time on the leaf-nodes (the todos). Now, you want to invoice your customer. If we don't show recursive times, then it won't be easy to know how many hours to bill your customer - it'll only be easy to know how much time you've spent on specific tasks. 

NB: Currently, there's no way in search to show the context of a header (which could be nice to have at some point, otherwise it's hard to make a distinction between headers that have the same name, but that's another matter). Since there's no context

**Comment by munen** [comment]
I have some points regarding the `clock:` filter:

1. I've found a potential bug (or just something that still would have to be implemented): When filtering on a date, I'm getting the same times for the header shown, no matter what 'matching' timestamp I enter.

2. Should the timestamps already support YYYYMM or YYYY-MM format next to YYYY? If so, then it's either not working or I'm doing something wrong.

Here's a demo showing both points. On the left, you can see the header with the logbook items for 2019 and 2020. On the right, I'm searching. Both searches for 2019 and 2020 return the same time and searching by month doesn't yield a result.

![question_on_dates](https://user-images.githubusercontent.com/505721/95905069-a6d37c80-0d98-11eb-9ff6-17eef60d125c.gif)

**Comment by tarnung** [comment]
about 1: there is nothing implemented yet. It just shows the recursive sum as it does in the regular header list. I agree that it should only take into account the matching clock entries though. Also it always shows the summed time if the setting is activated while i think in search it should only show it when a `clock:` searchterm is active That leads into the point about recursive sums in search. At the moment an item only shows up on a `clock:` search if it has time tracked on it. Not if it has time tracked on a child node. So my instinct was to display the found (presumably) leaf nodes and a sum of their time tracked. Getting grouped sums via parent headers is probably nicer than just displaying a sum over all matches (the two are not exclusive anyway). The thing I wouldn't like is my search results being leaf nodes with time logged on them mixed with parent nodes with time summed up. Right now you could not even tell one from the other. We could color code wether it's one or the other. Or we could show both time on item and recursive sum since a header can have both.

about 2: That worked before. I guess my efforts to simplify the parser did backfire in some cases. `clock:2020..2020` doesn't even parse anymore. There's a reason it's only a draft pr as of now.

**Comment by schoettl** [comment]
> > But in search it should probably show a simple sum of the time tracked on this header, not a recursive sum.
> 
> Curious, why do you say so? Let's say you're tracking time for a customer which has many projects (or epics) with many todos each.

Well, in org mode you would use clock tables to create invoices. It is a powerful concept and if you want good time reports you should really give org mode clock tables a chance. It's also easier to copy & paste or to convert the table to PDF/HTML and such things.

But I also would not make the search results too "clever". I think that each header should sum up only its own clock entries in the search results. Because the search results are a flat list and that should be reflected in the clock time, too, IMO. In your example, if you want to see the total time for a customer, you can probably easily navigate to the header in the header view (or just click on the search result).

**Comment by munen** [comment]
> Well, in org mode you would use clock tables to create invoices.

*hihi*, I know. I'm using them for controlling, invoicing and to automatically track time into customers time tracking systems. I have also brought them up in the original clock display PR from tarnung (https://github.com/200ok-ch/organice/pull/497#issuecomment-706586662).

Having said that, it's beside the point that Emacs Org mode has another feature for summing up clock entries. For reasons discussed in the original PR, we've went with clock display.

> But I also would not make the search results too "clever".

It's not for two reasons:

1. Emacs Org mode shows times recursively:

![image](https://user-images.githubusercontent.com/505721/95958675-c2bf3880-0e01-11eb-89eb-930d1fd47811.png)

2. The initial reason for adding the date filters was to omit implementing the more rigid `org-clock-display-default-range` (discussion here: https://github.com/200ok-ch/organice/pull/497#issuecomment-706574168 and following).

> In your example, if you want to see the total time for a customer, you can probably easily navigate to the header in the header view (or just click on the search result).

How would I see the time spend for a customer/project for a certain time period (i.e. the last month) like that? For a simplified customer project like:

```
* PROJECT Some customer project
** TODO Some task
   :LOGBOOK:
   CLOCK: [2020-09-13 Sun 09:35]--[2020-09-13 Sun 09:40] =>  0:05
   :END:
*** TODO Another task
    :LOGBOOK:
    CLOCK: [2020-10-14 Wed 09:35]--[2020-10-14 Wed 09:40] =>  0:05
    :END:
```

If I understand you correctly, when I search for clocked time for the last month like `clock:2020-09..`, I will:

1. Not find `PROJECT Some customer project`.
2. Even if I found it, and I would tap on it, I'd be back in the regular view where there is no scope for the time

Hence, my only option would be to manually add all the TODO headers times for the customer. Unless I misunderstand your proposal, that is.

**Comment by munen** [comment]
> about 1: there is nothing implemented yet.

👍

> It just shows the recursive sum as it does in the regular header list. I agree that it should only take into account the matching clock entries though.

👍

> Also it always shows the summed time if the setting is activated while i think in search it should show it when a clock: searchterm is active without regards wether the 'org-clock-display' setting is active.

I have no strong feelings either way as I can see your argument. Having said that, currently `org-clock-display` is a setting, so I would have assumed that it shows nonetheless. Personally, I'd opt to not add additional logic to the component. But if you have a strong feeling towards only showing the time in search when a `clock:` search is active, I'm not against ist.

> That leads into the point about recursive sums in search. At the moment an item only shows up on a clock: search if it has time tracked on it. Not if it has time tracked on a child node. So my instinct was to display the found (presumably) leaf nodes and a sum of their time tracked. Getting grouped sums via parent headers is probably nicer than just displaying a sum over all matches (the two are not exclusive anyway). The thing I wouldn't like is my search results being leaf nodes with time logged on them mixed with parent nodes with time summed up.

Tbh, I would not mind. That's how the Org agenda works, too.

> Right now you could not even tell one from the other.

I would assume you could - by reading the actual headline. Even though it's a flat list, it should be possible to find the relevant header either by typing out it's name or by having visible distinctions in the naming of the headers. I'd assume that a project/customer/epic header reads differently than a specific todo within said project/customer/epic.

> We could color code wether it's one or the other. Or we could show both time on item and recursive sum since a header can have both.

Showing both times is a good idea, I like it!

Ever since we've implemented search, I've been thinking on how to generally solve the problem of seeing the scope of the found header. In big Org files, I often encounter the following issue. Here's an example Org file:

```
* Foo
** Bar
* Foobar
** Bar
* Some header I want to refile into =Bar=
```

Now, let's say I search for `Bar`, then I cannot tell which is the right one.

The best solution I've come up with so far is to add breadcrumbs to outline the hierarchy. Something like this:

![image](https://user-images.githubusercontent.com/505721/95961495-68c07200-0e05-11eb-97c1-b2372c0c1970.png)


> about 2: That worked before. I guess my efforts to simplify the parser did backfire in some cases. clock:2020..2020 doesn't even parse anymore. 

Maybe a regression test will help. I'm actually quite impressed how far you've progressed in your PRs without any tests! But in any non-trivial system, it'll be easy to trip up something for any kind of change^^

> There's a reason it's only a draft pr as of now.

No worries, you've clearly stated that. I have no expectations of it being 'finished', I'm just trying to give good early feedback. My apologies if it came through too harsh. In any case, it wasn't meant like that, I'm quite impressed equally stoked on the feature set you're bringing in, the speed you're building them and the quality of the code you're delivering! :100:

**Comment by schoettl** [comment]
Alright, let's always show the recursive sum – in the search and in the header view. Making a difference between the two wouldn't be easy to understand.  @munen I didn't wanted to explain the clock table to you but rather to Martin ;)

**Comment by munen** [comment]
> Alright, let's always show the recursive sum – in the search and in the header view. Making a difference between the two wouldn't be easy to understand. 

:+1: 

> @munen I didn't wanted to explain the clock table to you but rather to Martin ;)

No worries; I've actually had this discussion with Martin in the last PR, too^^ And there, I've learnt that he's not an Emacs user, but has a different editor preference. However, he does like Org mode and recently changed his usage from Orgzly to organice(;

I actually like this progress in reach - I've talked to quite a few users by now who just use organice, but are not Emacs users. It's a trend that I'd like to foster. Of course, we'll always stay true to the Org format and general feature compatability, but if the UX and the feature set is good enough to use it standalone for project management for a non Emacs user, that would be a huge win and would tap into a bigger user pool. I'm super stoked to see that we now even have contributors from that pool :confetti_ball:

**Comment by munen** [comment]
As for clock tables, I really like them a lot. But they are quite coupled to Emacs, because it's easy to include a `formula` in elisp. So, even if we integrated them with the basic feature set, that would already be massively more complex. And to get them in completely, we'd have to have an elisp interpreter. And I think this is where we'll have a line for a very long time^^

That's why I'm super happy that Martin brings the clock display feature. It's less generic, but much easier to use and understand. It's more a thing for mobile/web apps whereas the clock report feels like a very Emacs-y feature to me (which personally, I like, of course. But not everyone shares that sentiment for several reasons.)

**Comment by tarnung** [comment]
Thanks for all the additional input.

@schoettl wrote
> But I also would not make the search results too "clever". I think that each header should sum up only its own clock entries in the search results. Because the search results are a flat list and that should be reflected in the clock time, too, IMO. In your example, if you want to see the total time for a customer, you can probably easily navigate to the header in the header view (or just click on the search result).

Upon reflection I agree. I was intrigued by the idea of making more data visible, but search probably isn't the place for that.

@schoettl wrote
> Alright, let's always show the recursive sum – in the search and in the header view. Making a difference between the two wouldn't be easy to understand.

I disagree. Mostly because I think the search timerange should be respected in the result shown. When only showing the recursive sum, It would be the sum over all time tracked without regard for the individual clock entries meeting the search criteria. 
I currently don't want the headache of calculating the recursive sum of all the clock entries on all the subheaders which meet all the search terms. (This would be necessary to show relevant parent headers anyway.)

> How would I see the time spend for a customer/project for a certain time period (i.e. the last month) like that? 

One way to more cleanly allow for that in the context of search is inherited tags (which are not yet something organice recognizes currently as far as I know). So right now you could do it if all your leaf nodes were tagged.

My plans for now are:
- make all of the search logic work

Let's keep this pr focussed on that feature.

Which means:
- keep the search logic as it is: on an header by header basis. No searching for any child clock entries or parent tags. 
- display the recursive sum of clock entries when the setting is enabled. Not because I like it that way but because that's what currently is implemented.

Next steps i'd consider in a pr after that:
- adjust how time clocked is shown in search results since how it is shown right now is only an unintentional byproduct of my efforts to display clock times in the header list.
- keep what is displayed in the search results tied to the individual header and the search term => show only time tracked on the header itself and only the sum of clock entries matching the search timerange.
- displaying the sum of all clock entries matching the search terms globally

**Comment by munen** [comment]
@tarnung wrote:

> Upon reflection I agree. I was intrigued by the idea of making more data visible, but search probably isn't the place for that.

So, if I understand you correctly, this means that your implementation will not be an alternative to implementing `org-clock-display-default-range`, correct? As I understood you in #497 in your comment https://github.com/200ok-ch/organice/pull/497#issuecomment-706583861, that was your original intention. 

> I disagree. Mostly because I think the search timerange should be respected in the result shown. When only showing the recursive sum, It would be the sum over all time tracked without regard for the individual clock entries meeting the search criteria.

Why? That it's currently doing that is just an implementation detail. You're currently calculating the sum for the total time in the `org` reducer. However, calculating it could also be part of the action `SET_SEARCH_FILTER_INFORMATION` which is triggered on changed search in put.

> One way to more cleanly allow for that in the context of search is inherited tags (which are not yet something organice recognizes currently as far as I know). So right now you could do it if all your leaf nodes were tagged.

You are correct that organice does not support inherited tags. However, I also fail to understand how tagging and `clock`ing relates to one another. I mean, I see one way where that might help - I could create a tag per customer/epic/time that's relevant to sum up. However, that would be a work-around (as would be creating separate headers to track for a customer/epic/time) and wouldn't show a summed up time on the main header since we're now only showing the time per headline. I'm assuming you have a different workflow in mind. Could you elaborate, please?

> Next steps i'd consider in a pr after that:
>  -  displaying the sum of all clock entries matching the search terms globally

That sounds like a feasible workaround. If we're not summing times on a header level, then the user can use multiple very specific search queries to get multiple lump sums (for example to get the time spent on `development`, `project management` and so on).

**Comment by tarnung** [comment]
> So, if I understand you correctly, this means that your implementation will not be an alternative to implementing `org-clock-display-default-range`, correct? As I understood you in #497 in your comment [#497 (comment)](https://github.com/200ok-ch/organice/pull/497#issuecomment-706583861), that was your original intention.

Yes it won't. I still think search is not the right place for it since it produces independent entries in a flat list. Search has a consistent behaviour right now and it would be strange if one search term breaks that.

> You're currently calculating the sum for the total time in the `org` reducer. However, calculating it could also be part of the action `SET_SEARCH_FILTER_INFORMATION` which is triggered on changed search in put.

That's what #520 does on a per-item basis. 

> > * displaying the sum of all clock entries matching the search terms globally
> 
> That sounds like a feasible workaround. If we're not summing times on a header level, then the user can use multiple very specific search queries to get multiple lump sums (for example to get the time spent on `development`, `project management` and so on).

I hope that it is enough of a workaround, at least combined with narrow/focus search.

**Comment by munen** [comment]
> Yes it won't. I still think search is not the right place for it since it produces independent entries in a flat list. Search has a consistent behaviour right now and it would be strange if one search term breaks that.

From a UX perspective, imo it would not break that behavior since clock entries recursively belong to all parents. That can be shown by using `org-clock-report` in Emacs for example. Here's a tiny subset of time tracking of one project. You can see in the generated table that the time entries bubble up:

![image](https://user-images.githubusercontent.com/505721/96686088-5c489600-137e-11eb-85f4-83e6e60d1caa.png)

It's the same behavior as `org-clock-display` if I create a sparse-tree in Emacs on the regexp "#operations". The second sub-heading "#dev" will not be visible. Note how "Timetracking" still shows "1:39" even though the visible headers ("#operations") only sum up to "0:30". 

Anyway, I just reiterate my standpoint, because you reiterated yours. 

If you're saying that it would break the consistent behavior from an implementation perspective, then I would understand. Because to mimic this behavior, you cannot just aggregate from the filtered headers, but would have to look at the complete list.


> I hope that it is enough of a workaround, at least combined with narrow/focus search.

If the feature works for you, that's good enough for me.

Personally, for my current projects, I would have to do quite a bit of narrowing/widening. Having said that, if I had used the functionality as we have it now/soon in organice, I would very likely have tracked the time differently, so all good(;

**Comment by tarnung** [comment]
Since your approach supports a wider variety of workflows we should at least try it. I'll give it a go. We'll see whether I understood your intent correctly and whether I get it to work. 
Then we can compare.

**Comment by tarnung** [comment]
I guess we can close this issue.

**Comment by munen** [comment]
Good point. 

Closed by #524.
